README.md
# Russian AI Cup (not only Russian!)
### Artificial intelligence programming contest. Official website: http://russianaicup.ru
Russian AI Cup — intitiative of the company Mail.Ru Group within the IT-oriented competitions.In this championship participants compete in skills of creating an artificial intelligence on example of game stratefies. Organizers are Mail.Ru Group and Codeforces. Best participants will receive special prizes.

Russian AI Cup — largest annual artificial intelligence programming contest in Russia, and third open competition for talented IT-specialists that is part of Mail.Ru Group strategy of forming and developing competitive Russian IT-industry on a global scale.

## RAIC 2018: CodeBall
We are pleased to welcome all the participants of the Russian AI Cup: CodeBall 2018 open beta!
The beta test will run until 21:00 UTC on December 23. Please note that at this time we can make a significant changes in the rules, scoring system and any other aspects of the championship. This week we will try to fix possible mistakes, optimize performance and make any other necessary improvements. The ratings will be reset after this week of beta.
Some links on the website are not working. We will fix it within a few days.
Good luck and have fun!

## UPDATE 19.12.18
[Here](https://github.com/MailRuChamps/raic/blob/master/UPDATE-19.12.md) you can read about updates.

## Useful links
1. About [CodeBall](http://russianaicup.ru). News, Notifications;
2. Quick start [instructions](http://russianaicup.ru/p/quick). Other sources will be published soon on this repository and [here](http://russianaicup.ru/p/examples);
3. Play CodeBall in the [browser](http://russianaicup.ru/play).

## Community for English speakers (Discord)
We are [HERE](https://discord.gg/yhJRg7D). Join us! 

## Community for Russian speakers (Telegram)
[HERE](https://t.me/aicups). Join us! 

## Official Contacts
Also, we've email **cups@corp.mail.ru**. If you have any private questions, you can ask us. We always check our inbox and reply to all.


CHANGELOG.md
# 1.1.1

- Added `outside opacity` setting
- Camera & rendering mode is now saved
- Fixed custom rendered sphere sizes

# 1.0.1

- Added `--disable-random` option (uses `MAX_HIT_E` for collisions, robots sorted by `id`)


QUICKSTARTS/с++/StarterC++.h
#if defined(_MSC_VER) && (_MSC_VER >= 1200)
#pragma once
#endif

#ifndef _MY_STRATEGY_H_
#define _MY_STRATEGY_H_

#include "Strategy.h"
#include "math.h"

const double EPS = 1e-5;

const double BALL_RADIUS = 2.0;
const double ROBOT_MAX_RADIUS = 1.05;
const double MAX_ENTITY_SPEED = 100.0;
const double ROBOT_MAX_GROUND_SPEED = 30.0;
const double ROBOT_MAX_JUMP_SPEED = 15.0;

const double JUMP_TIME = 0.2;
const double MAX_JUMP_HEIGHT = 3.0;

struct Point2D {
    double x {0.0};
    double z {0.0};
    Point2D() {}
    Point2D(double x, double z) : x(x), z(z) {}
    void set(double x_, double z_) { x = x_; z = z_; }
    double dist() { return sqrt(x*x + z*z); }
    Point2D normalize(double len) { return { x/len, z/len }; }
    double distTo(double x_, double z_) { return sqrt((x-x_)*(x-x_) + (z-z_)*(z-z_)); }
    double distTo(Point2D p) { return distTo(p.x, p.z); }
    Point2D operator+(Point2D p) { return { x+p.x, z+p.z }; }
    Point2D operator*(double val) { return { x*val, z*val }; }
    void operator*=(double val) { x *= val; z *= val; }
};

struct Point3D : Point2D {
    double y {0.0}; // высота
    Point3D() {}
    Point3D(double x, double z, double y) : y(y) { Point2D::set(x, z); }
    void set(double x_, double z_, double y_) { x = x_; z = z_; y = y_; }
    double dist() { return sqrt(x*x + y*y + z*z); }
    double distTo(double x_, double z_, double y_) { return sqrt((x-x_)*(x-x_) + (y-y_)*(y-y_) + (z-z_)*(z-z_)); }
    double distTo(Point3D p) { return distTo(p.x, p.z, p.y); }
    Point3D operator+(Point3D p) { return { x+p.x, z+p.z, y+p.y };}
    Point3D operator*(double val) { return { x*val, z*val, y*val }; }
    void operator*=(double val) { x *= val; z *= val; y *= val; }
};

class MyStrategy : public Strategy {
public:

    Point3D ball;
    Point3D ball_v;

    MyStrategy();

    void act(const model::Robot& me, const model::Rules& rules, const model::Game& world, model::Action& action) override;

};

#endif


QUICKSTARTS/с++/StarterC++.cpp
#include "MyStrategy.h"

using namespace model;

MyStrategy::MyStrategy() { }

void MyStrategy::act(const Robot& me, const Rules& rules, const Game& game, Action& action) {

    ball.set(game.ball.x, game.ball.z, game.ball.y);
    ball_v.set(game.ball.velocity_x, game.ball.velocity_z, game.ball.velocity_y);

    // Наша стратегия умеет играть только на земле
    // Поэтому, если мы не касаемся земли, будет использовать нитро
    // чтобы как можно быстрее попасть обратно на землю
    if ( !me.touch ) {
        action.target_velocity_x = 0.0;
        action.target_velocity_z = 0.0;
        action.target_velocity_y = -MAX_ENTITY_SPEED;
        action.jump_speed        = 0.0;
        action.use_nitro         = true;
        return;
    }

    // Если при прыжке произойдет столкновение с мячом, и мы находимся
    // с той же стороны от мяча, что и наши ворота, прыгнем, тем самым
    // ударив по мячу сильнее в сторону противника
    bool jump = (   ball.distTo(me.x, me.z, me.y) < (BALL_RADIUS + ROBOT_MAX_RADIUS)
                 && me.z < ball.z );

    // Так как роботов несколько, определим нашу роль - защитник, или нападающий
    // Нападающим будем в том случае, если есть дружественный робот,
    // находящийся ближе к нашим воротам
    bool is_attacker = false; // = (game.robots.size() == 2);
    for (const Robot &robot : game.robots) {
        if (   robot.is_teammate
            && robot.id != me.id) {
            if (robot.z < me.z) {
                is_attacker = true;
            }
        }
    }

    if (is_attacker) {
        // Стратегия нападающего:
        // Просимулирем примерное положение мяча в следующие 10 секунд, с точностью 0.1 секунда
        for (int i = 1; i <= 100; ++i) {
            double t = i * 0.1;
            Point3D ball_pos = ball + ball_v*t;
            // Если мяч не вылетит за пределы арены
            // (произойдет столкновение со стеной, которое мы не рассматриваем),
            // и при этом мяч будет находится ближе к вражеским воротам, чем робот,
            if (   ball_pos.z > me.z
                && abs(ball_pos.x) < (rules.arena.width / 2.0)
                && abs(ball_pos.z) < (rules.arena.depth / 2.0) ) {
                // Посчитаем, с какой скоростью робот должен бежать,
                // Чтобы прийти туда же, где будет мяч, в то же самое время
                Point2D delta_pos(ball_pos.x - me.x, ball_pos.z - me.z);
                double delta_pos_dist = delta_pos.dist();
                double need_speed = delta_pos_dist / t;
                // Если эта скорость лежит в допустимом отрезке
                if (0.5 * ROBOT_MAX_GROUND_SPEED < need_speed
                    && need_speed < ROBOT_MAX_GROUND_SPEED ) {
                    // То это и будет наше текущее действие
                    Point2D target_velocity(delta_pos.normalize(delta_pos_dist)*need_speed);
                    action.target_velocity_x = target_velocity.x;
                    action.target_velocity_z = target_velocity.z;
                    action.target_velocity_y = 0.0;
                    action.jump_speed = jump ? ROBOT_MAX_JUMP_SPEED : 0.0;
                    action.use_nitro = false;
                    return;
                }
            }
        }
    }
    // Стратегия защитника (или атакующего, не нашедшего хорошего момента для удара):
    // Будем стоять посередине наших ворот
    Point2D target_pos(0.0, -(rules.arena.depth / 2.0) + rules.arena.bottom_radius);
    // Причем, если мяч движется в сторону наших ворот
    if (ball_v.z < -EPS) {
        // Найдем время и место, в котором мяч пересечет линию ворот
        double t = (target_pos.z - ball.z) / ball_v.z;
        double x = ball.x + ball_v.x * t;
        // Если это место - внутри ворот
        if (abs(x) < (rules.arena.goal_width / 2.0)) {
            // То пойдем защищать его
            target_pos.x = x;
        }
    }
    // Установка нужных полей для желаемого действия
    Point2D target_velocity(target_pos.x - me.x, target_pos.z - me.z);
    target_velocity *= ROBOT_MAX_GROUND_SPEED;

    action.target_velocity_x = target_velocity.x;
    action.target_velocity_z = target_velocity.z;
    action.target_velocity_y = 0.0;
    action.jump_speed = jump ? ROBOT_MAX_JUMP_SPEED : 0.0;
    action.use_nitro = false;
}


QUICKSTARTS/csharp/MyStrategy.cs
using System;
using Com.CodeGame.CodeBall2018.DevKit.CSharpCgdk.Model;
using Action = Com.CodeGame.CodeBall2018.DevKit.CSharpCgdk.Model.Action;

namespace Com.CodeGame.CodeBall2018.DevKit.CSharpCgdk
{
    // Логика стратегии скопирована из базовой стратегии c++ (SannikovDmitry)
    public sealed class MyStrategy : IStrategy
    {
        private const double EPS = 1e-5;
        
        public void Act(Robot me, Rules rules, Game game, Action action)
        {
            var ball = game.ball;

            // Наша стратегия умеет играть только на земле
            // Поэтому, если мы не касаемся земли, будет использовать нитро
            // чтобы как можно быстрее попасть обратно на землю
            if (!me.touch)
            {
                action.target_velocity_x = 0.0;
                action.target_velocity_z = 0.0;
                action.target_velocity_y = -rules.MAX_ENTITY_SPEED;
                action.jump_speed = 0.0;
                action.use_nitro = true;
                return;
            }

            // Если при прыжке произойдет столкновение с мячом, и мы находимся
            // с той же стороны от мяча, что и наши ворота, прыгнем, тем самым
            // ударив по мячу сильнее в сторону противника
            bool jump = (DistBetweenBallAndRobot3D(ball, me) < (rules.BALL_RADIUS + rules.ROBOT_MAX_RADIUS) &&
                         me.y < ball.y);

            // Так как роботов несколько, определим нашу роль - защитник, или нападающий
            // Нападающим будем в том случае, если есть дружественный робот,
            // находящийся ближе к нашим воротам
            bool is_attacker = false;
            foreach (var robot in game.robots)
            {
                if (robot.is_teammate && robot.id != me.id)
                {
                    if (robot.z < me.z)
                    {
                        is_attacker = true;
                    }
                }
            }

            if (is_attacker)
            {
                // Стратегия нападающего:
                // Просимулирем примерное положение мяча в следующие 10 секунд, с точностью 0.1 секунда
                for (int i = 0; i < 100; ++i)
                {
                    double t = i * 0.1;
                    Ball ball_pos = CalcNewBallPos(ball, t);
                    // Если мяч не вылетит за пределы арены
                    // (произойдет столкновение со стеной, которое мы не рассматриваем),
                    // и при этом мяч будет находится ближе к вражеским воротам, чем робот,
                    if (ball_pos.z > me.z
                        && Math.Abs(ball.x) < (rules.arena.width / 2.0)
                        && Math.Abs(ball.z) < (rules.arena.depth / 2.0))
                    {
                        // Посчитаем, с какой скоростью робот должен бежать,
                        // Чтобы прийти туда же, где будет мяч, в то же самое время
                        double delta_pos_x = ball_pos.x - me.x;
                        double delta_pos_z = ball_pos.z - me.z;
                        double delta_pos_dist = Math.Sqrt(delta_pos_x*delta_pos_x + delta_pos_z*delta_pos_z);
                        double need_speed = delta_pos_dist / t;
                        // Если эта скорость лежит в допустимом отрезке
                        if (0.5 * rules.ROBOT_MAX_GROUND_SPEED < need_speed
                            && need_speed < rules.ROBOT_MAX_GROUND_SPEED)
                        {
                            // То это и будет наше текущее действие
                            action.target_velocity_x = delta_pos_x / delta_pos_dist * need_speed;
                            action.target_velocity_z = delta_pos_z / delta_pos_dist * need_speed;
                            action.target_velocity_y = 0.0;
                            action.jump_speed = jump ? rules.ROBOT_MAX_JUMP_SPEED : 0.0;
                            action.use_nitro = false;
                            return;
                        }
                    }
                }
            }
            
            // Стратегия защитника (или атакующего, не нашедшего хорошего момента для удара):
            // Будем стоять посередине наших ворот
            double target_pos_x = 0.0;
            double target_pos_z = -(rules.arena.depth / 2.0) + rules.arena.bottom_radius;
            // Причем, если мяч движется в сторону наших ворот
            if (ball.velocity_z < -EPS) {
                // Найдем время и место, в котором мяч пересечет линию ворот
                double t = (target_pos_z - ball.z) / ball.velocity_z;
                double x = ball.x + ball.velocity_x * t;
                // Если это место - внутри ворот
                if (Math.Abs(x) < (rules.arena.goal_width / 2.0)) {
                    // То пойдем защищать его
                    target_pos_x = x;
                }
            }
            
            // Установка нужных полей для желаемого действия
            double target_velocity_x = (target_pos_x - me.x)*rules.ROBOT_MAX_GROUND_SPEED;
            double target_velocity_z = (target_pos_z - me.z)*rules.ROBOT_MAX_GROUND_SPEED;
            
            action.target_velocity_x = target_velocity_x;
            action.target_velocity_z = target_velocity_z;
            action.target_velocity_y = 0.0;
            action.jump_speed = jump ? rules.ROBOT_MAX_JUMP_SPEED : 0.0;
            action.use_nitro = false;
        }

        public string CustomRendering()
        {
            return "";
        }
        
        private static double DistBetweenBallAndRobot3D(Ball b, Robot r)
        {
            return Math.Sqrt((b.x - r.x) * (b.x - r.x) + (b.y - r.y) * (b.y - r.y) + (b.z - r.z) * (b.z - r.z));
        }

        private static double DistBetweenBallAndRobot2D(Ball b, Robot r)
        {
            return Math.Sqrt((b.x - r.x) * (b.x - r.x) + (b.z - r.z) * (b.z - r.z));
        }

        private static Ball CalcNewBallPos(Ball b, double t)
        {
            return new Ball() {x = b.x + b.velocity_x * t, y = b.y + b.velocity_y * t, z = b.z + b.velocity_z * t};
        }
    }
}


QUICKSTARTS/python3/MyStrategy.py
from model.action import Action
from model.game import Game
from model.robot import Robot
from model.rules import Rules

EPS = 1e-5
# Константы, взятые из документации
BALL_RADIUS = 2.0
ROBOT_MAX_RADIUS = 1.05
MAX_ENTITY_SPEED = 100.0
ROBOT_MAX_GROUND_SPEED = 30.0
ROBOT_MAX_JUMP_SPEED = 15.0

JUMP_TIME = 0.2
MAX_JUMP_HEIGHT = 3.0


class Vector2D:
    # Нам понадобится работа с 2d векторами
    def __init__(self, x=0.0, z=0.0):
        self.x = x
        self.z = z

    # Нахождение длины вектора
    def len(self):
        return ((self.x * self.x) + (self.z * self.z))**0.5

    # Операция - для векторов
    def __sub__(self, other):
        return Vector2D(self.x - other.x, self.z - other.z)

    # Операция + для векторов
    def __add__(self, other):
        return Vector2D(self.x + other.x, self.z + other.z)

    # Операция умножения вектора на число
    def __mul__(self, num: float):
        return Vector2D(self.x * num, self.z * num)

    # Нормализация вектора (приведение длины к 1)
    def normalize(self):
        return Vector2D(self.x/self.len(), self.z/self.len())


class MyStrategy:
    # Код стратегии
    def act(self, me: Robot, rules: Rules, game: Game, action: Action):

        # Наша стратегия умеет играть только на земле
        # Поэтому, если мы не касаемся земли, будет использовать нитро
        # чтобы как можно быстрее попасть обратно на землю
        if not me.touch:
            action.target_velocity_x = 0.0
            action.target_velocity_y = -MAX_ENTITY_SPEED
            action.target_velocity_z = 0.0
            action.jump_speed = 0.0
            action.use_nitro = True
            return

        dist_to_ball = ((me.x - game.ball.x) ** 2
                        + (me.y - game.ball.y) ** 2
                        + (me.z - game.ball.z) ** 2
                        ) ** 0.5

        # Если при прыжке произойдет столкновение с мячом, и мы находимся
        # с той же стороны от мяча, что и наши ворота, прыгнем, тем самым
        # ударив по мячу сильнее в сторону противника
        jump = (dist_to_ball < BALL_RADIUS +
                ROBOT_MAX_RADIUS and me.z < game.ball.z)

        # Так как роботов несколько, определим нашу роль - защитник, или нападающий
        # Нападающим будем в том случае, если есть дружественный робот,
        # находящийся ближе к нашим воротам
        is_attacker = len(game.robots) == 2

        for robot in game.robots:
            robot: Robot = robot
            if robot.is_teammate and robot.id != me.id:
                if robot.z < me.z:
                    is_attacker = True

        if is_attacker:
            # Стратегия нападающего:
            # Просимулирем примерное положение мяча в следующие 10 секунд, с точностью 0.1 секунда
            for i in range(1, 101):
                t = i * 0.1
                ball_x = game.ball.x
                ball_z = game.ball.z
                ball_vel_x = game.ball.velocity_x
                ball_vel_z = game.ball.velocity_z
                ball_pos = Vector2D(ball_x, ball_z) + \
                    Vector2D(ball_vel_x, ball_vel_z) * t

                # Если мяч не вылетит за пределы арены
                # (произойдет столкновение со стеной, которое мы не рассматриваем),
                # и при этом мяч будет находится ближе к вражеским воротам, чем робот,
                if ball_pos.z > me.z \
                        and abs(ball_pos.x) < (rules.arena.width / 2.0) \
                        and abs(ball_pos.z) < (rules.arena.depth / 2.0):

                    # Посчитаем, с какой скоростью робот должен бежать,
                    # Чтобы прийти туда же, где будет мяч, в то же самое время
                    delta_pos = Vector2D(
                        ball_pos.x, ball_pos.z) - Vector2D(me.x, me.z)

                    need_speed = delta_pos.len() / t
                    # Если эта скорость лежит в допустимом отрезке
                    if 0.5 * ROBOT_MAX_GROUND_SPEED < need_speed \
                            and need_speed < ROBOT_MAX_GROUND_SPEED:
                        # То это и будет наше текущее действие
                        target_velocity = delta_pos.normalize() * need_speed
                        action.target_velocity_x = target_velocity.x
                        action.target_velocity_y = 0.0
                        action.target_velocity_z = target_velocity.z
                        action.jump_speed = ROBOT_MAX_JUMP_SPEED if jump else 0.0
                        action.use_nitro = False
                        return

        # Стратегия защитника (или атакующего, не нашедшего хорошего момента для удара):
        # Будем стоять посередине наших ворот
        target_pos = Vector2D(
            0.0, -(rules.arena.depth / 2.0) + rules.arena.bottom_radius)
        # Причем, если мяч движется в сторону наших ворот
        if game.ball.velocity_z < -EPS:
            # Найдем время и место, в котором мяч пересечет линию ворот
            t = (target_pos.z - game.ball.z) / game.ball.velocity_z
            x = game.ball.x + game.ball.velocity_x * t
            # Если это место - внутри ворот
            if abs(x) < rules.arena.goal_width / 2.0:
                # То пойдем защищать его
                target_pos.x = x
        # Установка нужных полей для желаемого действия
        target_velocity = Vector2D(
            target_pos.x - me.x, target_pos.z - me.z) * ROBOT_MAX_GROUND_SPEED

        action.target_velocity_x = target_velocity.x
        action.target_velocity_y = 0.0
        action.target_velocity_z = target_velocity.z
        action.jump_speed = ROBOT_MAX_JUMP_SPEED if jump else 0.0
        action.use_nitro = False


QUICKSTARTS/rust/MyStrategy.rs
use model::*;
use strategy::Strategy;

const EPS: f64 = 1e-5;

// Константы, взятые из документации
const BALL_RADIUS: f64 = 2.0;
const ROBOT_MAX_RADIUS: f64 = 1.05;
const MAX_ENTITY_SPEED: f64 = 100.0;
const ROBOT_MAX_GROUND_SPEED: f64 = 30.0;
const ROBOT_MAX_JUMP_SPEED: f64 = 15.0;

#[derive(Default)]
pub struct MyStrategy;

// Нам понадобится работа с 2d векторами
#[derive(Copy, Clone, Debug)]
struct Vec2 {
    x: f64,
    y: f64,
}

impl Vec2 {
    fn new(x: f64, y: f64) -> Self {
        Self { x, y }
    }
    // Нахождение длины вектора
    fn len(&self) -> f64 {
        (self.x * self.x + self.y * self.y).sqrt()
    }
    // Нормализация вектора (приведение длины к 1)
    fn normalize(self) -> Self {
        self * (1.0 / self.len())
    }
}

// Операция - для векторов
impl std::ops::Sub for Vec2 {
    type Output = Self;
    fn sub(self, b: Self) -> Self {
        Self::new(self.x - b.x, self.y - b.y)
    }
}

// Операция + для векторов
impl std::ops::Add for Vec2 {
    type Output = Self;
    fn add(self, b: Self) -> Self {
        Self::new(self.x + b.x, self.y + b.y)
    }
}

// Операция умножения вектора на число
impl std::ops::Mul<f64> for Vec2 {
    type Output = Self;
    fn mul(self, k: f64) -> Self {
        Self::new(self.x * k, self.y * k)
    }
}

// Определение общего интерфейса для робота и мяча
trait Entity {
    // Нахождение позиции объекта в плоскости XZ
    fn position(&self) -> Vec2;
    // Нахождение скорости объекта в плоскости XZ
    fn velocity(&self) -> Vec2;
}

// Реализация интерфейса Entity для робота
impl Entity for Robot {
    fn position(&self) -> Vec2 {
        Vec2::new(self.x, self.z)
    }
    fn velocity(&self) -> Vec2 {
        Vec2::new(self.velocity_x, self.velocity_z)
    }
}

// Реализация интерфейса Entity для мяча
impl Entity for Ball {
    fn position(&self) -> Vec2 {
        Vec2::new(self.x, self.z)
    }
    fn velocity(&self) -> Vec2 {
        Vec2::new(self.velocity_x, self.velocity_z)
    }
}

// Код стратегии
impl Strategy for MyStrategy {
    fn act(&mut self, me: &Robot, rules: &Rules, game: &Game, action: &mut Action) {

        // Наша стратегия умеет играть только на земле
        // Поэтому, если мы не касаемся земли, будет использовать нитро
        // чтобы как можно быстрее попасть обратно на землю
        if !me.touch {
            *action = Action {
                target_velocity_x: 0.0,
                target_velocity_y: -MAX_ENTITY_SPEED,
                target_velocity_z: 0.0,
                jump_speed: 0.0,
                use_nitro: true,
            };
            return;
        }

        const JUMP_TIME: f64 = 0.2;
        const MAX_JUMP_HEIGHT: f64 = 3.0;

        // Если при прыжке произойдет столкновение с мячом, и мы находимся
        // с той же стороны от мяча, что и наши ворота, прыгнем, тем самым
        // ударив по мячу сильнее в сторону противника
        let mut jump = ((me.x - game.ball.x).powi(2)
            + (me.y - game.ball.y).powi(2)
            + (me.z - game.ball.z).powi(2))
        .sqrt()
            < BALL_RADIUS + ROBOT_MAX_RADIUS
            && me.z < game.ball.z;

        // Так как роботов несколько, определим нашу роль - защитник, или нападающий
        // Нападающим будем в том случае, если есть дружественный робот,
        // находящийся ближе к нашим воротам
        let mut is_attacker = game.robots.len() == 2;
        for robot in &game.robots {
            if robot.is_teammate && robot.id != me.id {
                if robot.position().y < me.position().y {
                    is_attacker = true;
                }
            }
        }

        if is_attacker {
            // Стратегия нападающего:
            // Просимулирем примерное положение мяча в следующие 10 секунд, с точностью 0.1 секунда
            for i in 1..100 {
                let t = i as f64 * 0.1;
                let ball_pos = game.ball.position() + game.ball.velocity() * t;
                // Если мяч не вылетит за пределы арены
                // (произойдет столкновение со стеной, которое мы не рассматриваем),
                // и при этом мяч будет находится ближе к вражеским воротам, чем робот,
                if ball_pos.y > me.position().y
                    && ball_pos.x.abs() < (rules.arena.width / 2.0)
                    && ball_pos.y.abs() < (rules.arena.depth / 2.0)
                {
                    // Посчитаем, с какой скоростью робот должен бежать,
                    // Чтобы прийти туда же, где будет мяч, в то же самое время
                    let delta_pos = Vec2::new(ball_pos.x, ball_pos.y)
                        - Vec2::new(me.position().x, me.position().y);
                    let need_speed = delta_pos.len() / t;
                    // Если эта скорость лежит в допустимом отрезке
                    if 0.5 * ROBOT_MAX_GROUND_SPEED < need_speed
                        && need_speed < ROBOT_MAX_GROUND_SPEED
                    {
                        // То это и будет наше текущее действие
                        let target_velocity =
                            Vec2::new(delta_pos.x, delta_pos.y).normalize() * need_speed;
                        *action = Action {
                            target_velocity_x: target_velocity.x,
                            target_velocity_y: 0.0,
                            target_velocity_z: target_velocity.y,
                            jump_speed: if jump { ROBOT_MAX_JUMP_SPEED } else { 0.0 },
                            use_nitro: false,
                        };
                        return;
                    }
                }
            }
        }
        // Стратегия защитника (или атакующего, не нашедшего хорошего момента для удара):
        // Будем стоять посередине наших ворот
        let mut target_pos = Vec2::new(0.0, -(rules.arena.depth / 2.0) + rules.arena.bottom_radius);
        // Причем, если мяч движется в сторону наших ворот
        if game.ball.velocity().y < -EPS {
            // Найдем время и место, в котором мяч пересечет линию ворот
            let t = (target_pos.y - game.ball.position().y) / game.ball.velocity().y;
            let x = game.ball.position().x + game.ball.velocity().x * t;
            // Если это место - внутри ворот
            if x.abs() < (rules.arena.goal_width / 2.0) {
                // То пойдем защищать его
                target_pos.x = x;
            }
        }

        // Установка нужных полей для желаемого действия
        let target_velocity = Vec2::new(
            target_pos.x - me.position().x,
            target_pos.y - me.position().y,
        ) * ROBOT_MAX_GROUND_SPEED;
        *action = Action {
            target_velocity_x: target_velocity.x,
            target_velocity_y: 0.0,
            target_velocity_z: target_velocity.y,
            jump_speed: if jump { ROBOT_MAX_JUMP_SPEED } else { 0.0 },
            use_nitro: false,
        }
    }
}